<a href="https://colab.research.google.com/github/danishammar123/SALES/blob/main/dsl_cars_dataset.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [21]:
import pandas as pd

In [53]:
cars = pd.read_csv('/content/drive/MyDrive/2. Cars Data1.csv')
#when you paste the file path in the parenthesis remember to add single or double quotes marks('')

In [52]:
cars.head()

Make           Model   Type Origin DriveTrain      MSRP   Invoice  \
0  Acura             MDX    SUV   Asia        All  $36,945   $33,337    
1  Acura  RSX Type S 2dr  Sedan   Asia      Front  $23,820   $21,761    
2  Acura         TSX 4dr  Sedan   Asia      Front  $26,990   $24,647    
3  Acura          TL 4dr  Sedan   Asia      Front  $33,195   $30,299    
4  Acura      3.5 RL 4dr  Sedan   Asia      Front  $43,755   $39,014    

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         3.5        6.0       265.0      17.0         23.0  4451.0   
1         2.0        4.0       200.0      24.0         31.0  2778.0   
2         2.4        4.0       200.0      22.0         29.0  3230.0   
3         3.2        6.0       270.0      20.0         28.0  3575.0   
4         3.5        6.0       225.0      18.0         24.0  3880.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0  
2      105.0   183.0  
3      108.0   186.0  
4      115.0   197.0

In [35]:
cars.shape

(432, 15)

**CLEANING OF DATA:**

**Q1: FIND nULL VALES AND FILL THEM WITH MEAAN OF THAT COLUMN.**

In [54]:
#finding null vales
cars.isnull().sum()

Make           4
Model          4
Type           4
Origin         4
DriveTrain     4
MSRP           4
Invoice        4
EngineSize     4
Cylinders      6
Horsepower     4
MPG_City       4
MPG_Highway    4
Weight         4
Wheelbase      4
Length         4
dtype: int64

In [55]:
cars['Cylinders'].fillna(cars['Cylinders'].mean(), inplace=True)

In [60]:
cars.isnull().sum()

Make           0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     0
Cylinders      0
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

In [56]:
column_types = cars.dtypes
print(column_types)

Make            object
Model           object
Type            object
Origin          object
DriveTrain      object
MSRP            object
Invoice         object
EngineSize     float64
Cylinders      float64
Horsepower     float64
MPG_City       float64
MPG_Highway    float64
Weight         float64
Wheelbase      float64
Length         float64
dtype: object


In [57]:
cars.dropna(subset=['Make'], inplace=True)

In [58]:
cars.isnull().sum()

Make           0
Model          0
Type           0
Origin         0
DriveTrain     0
MSRP           0
Invoice        0
EngineSize     0
Cylinders      0
Horsepower     0
MPG_City       0
MPG_Highway    0
Weight         0
Wheelbase      0
Length         0
dtype: int64

In [44]:
cars.shape

(426, 15)

**Q2:(based on count value)**

what are the diffrent types of make are there in the dataset, and what is the count(occurance)of each make in the data?

In [59]:
cars['Make'].value_counts()

Toyota           28
Chevrolet        27
Mercedes-Benz    26
Ford             23
BMW              20
Audi             19
Honda            17
Nissan           17
Volkswagen       15
Chrysler         15
Dodge            13
Mitsubishi       13
Volvo            12
Jaguar           12
Hyundai          12
Subaru           11
Pontiac          11
Mazda            11
Lexus            11
Kia              11
Buick             9
Mercury           9
Lincoln           9
Saturn            8
Cadillac          8
Suzuki            8
Infiniti          8
GMC               8
Acura             7
Porsche           7
Saab              7
Land Rover        3
Oldsmobile        3
Jeep              3
Scion             2
Isuzu             2
MINI              2
Hummer            1
Name: Make, dtype: int64

**Q3:show all the records where origin is Asia or Europe**
(use  filtering method)

In [61]:
cars[cars['Origin'].isin(['Asia', 'Europe'])]

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      17.0         23.0   
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[281 rows x 15 columns]

**Q4:(Remove unwanted records**)
remove all records(rows) where weight is above 4000

In [62]:
cars['Weight'] > 4000

0      False
1      False
2      False
3      False
4      False
       ...  
427    False
428    False
429    False
430    False
431    False
Name: Weight, Length: 428, dtype: bool

In [63]:
#we dont use above syntex becuase we dont need our result in boolean format
#so we use following syntex
cars[cars['Weight'] > 4000]

Make                           Model    Type  Origin DriveTrain  \
0         Acura                             MDX     SUV    Asia        All   
15         Audi  A4 3.0 Quattro convertible 2dr   Sedan  Europe        All   
17         Audi              A6 4.2 Quattro 4dr   Sedan  Europe        All   
18         Audi                A8 L Quattro 4dr   Sedan  Europe        All   
20         Audi                        RS 6 4dr  Sports  Europe      Front   
..          ...                             ...     ...     ...        ...   
405  Volkswagen                      Touareg V6     SUV  Europe        All   
415  Volkswagen                     Phaeton 4dr   Sedan  Europe      Front   
416  Volkswagen                 Phaeton W12 4dr   Sedan  Europe      Front   
419  Volkswagen                       Passat W8   Wagon  Europe      Front   
420       Volvo                         XC90 T6     SUV  Europe        All   

         MSRP   Invoice  EngineSize  Cylinders  Horsepower  MPG_City  \
0    $36,945   $33,337          3.5        6.0       265.0      17.0   
15   $44,240   $40,075          3.0        6.0       220.0      18.0   
17   $49,690   $44,936          4.2        8.0       300.0      17.0   
18   $69,190   $64,740          4.2        8.0       330.0      17.0   
20   $84,600   $76,417          4.2        8.0       450.0      15.0   
..        ...       ...         ...        ...         ...       ...   
405  $35,515   $32,243          3.2        6.0       220.0      15.0   
415  $65,000   $59,912          4.2        8.0       335.0      16.0   
416  $75,000   $69,130          6.0       12.0       420.0      12.0   
419  $40,235   $36,956          4.0        8.0       270.0      18.0   
420  $41,250   $38,851          2.9        6.0       268.0      15.0   

     MPG_Highway  Weight  Wheelbase  Length  
0           23.0  4451.0      106.0   189.0  
15          25.0  4013.0      105.0   180.0  
17          24.0  4024.0      109.0   193.0  
18          24.0  4399.0      121.0   204.0  
20          22.0  4024.0      109.0   191.0  
..           ...     ...        ...     ...  
405         20.0  5086.0      112.0   187.0  
415         22.0  5194.0      118.0   204.0  
416         19.0  5399.0      118.0   204.0  
419         25.0  4067.0      106.0   184.0  
420         20.0  4638.0      113.0   189.0  

[103 rows x 15 columns]

In [64]:
#for removing above 4000 we use condition method ~
cars[~(cars['Weight'] > 4000)]

Make                    Model   Type  Origin DriveTrain      MSRP  \
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
5    Acura  3.5 RL w/Navigation 4dr  Sedan    Asia      Front  $46,100    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
1    $21,761          2.0        4.0       200.0      24.0         31.0   
2    $24,647          2.4        4.0       200.0      22.0         29.0   
3    $30,299          3.2        6.0       270.0      20.0         28.0   
4    $39,014          3.5        6.0       225.0      18.0         24.0   
5    $41,100          3.5        6.0       225.0      18.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      21.0         28.0   
428  $40,083          2.3        5.0       242.0      20.0         26.0   
429  $42,573          2.9        6.0       268.0      19.0         26.0   
430  $24,641          1.9        4.0       170.0      22.0         29.0   
431  $33,112          2.5        5.0       208.0      20.0         27.0   

     Weight  Wheelbase  Length  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
5    3893.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[325 rows x 15 columns]

note:you can see above that now we have 325 rows , originally we have 426 rows

**Q5:instruction (applying function on a column)**
>increase all the values of MPG_city column by 3

In [67]:
cars.head(2)

Make           Model   Type Origin DriveTrain      MSRP   Invoice  \
0  Acura             MDX    SUV   Asia        All  $36,945   $33,337    
1  Acura  RSX Type S 2dr  Sedan   Asia      Front  $23,820   $21,761    

   EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  Weight  \
0         3.5        6.0       265.0      20.0         23.0  4451.0   
1         2.0        4.0       200.0      27.0         31.0  2778.0   

   Wheelbase  Length  
0      106.0   189.0  
1      101.0   172.0

In [65]:
cars['MPG_City'] = cars['MPG_City'].apply(lambda x:x+3)

In [66]:
#check dataframe
cars

Make                    Model   Type  Origin DriveTrain      MSRP  \
0    Acura                      MDX    SUV    Asia        All  $36,945    
1    Acura           RSX Type S 2dr  Sedan    Asia      Front  $23,820    
2    Acura                  TSX 4dr  Sedan    Asia      Front  $26,990    
3    Acura                   TL 4dr  Sedan    Asia      Front  $33,195    
4    Acura               3.5 RL 4dr  Sedan    Asia      Front  $43,755    
..     ...                      ...    ...     ...        ...       ...   
427  Volvo  C70 LPT convertible 2dr  Sedan  Europe      Front  $40,565    
428  Volvo  C70 HPT convertible 2dr  Sedan  Europe      Front  $42,565    
429  Volvo               S80 T6 4dr  Sedan  Europe      Front  $45,210    
430  Volvo                      V40  Wagon  Europe      Front  $26,135    
431  Volvo                     XC70  Wagon  Europe        All  $35,145    

      Invoice  EngineSize  Cylinders  Horsepower  MPG_City  MPG_Highway  \
0    $33,337          3.5        6.0       265.0      20.0         23.0   
1    $21,761          2.0        4.0       200.0      27.0         31.0   
2    $24,647          2.4        4.0       200.0      25.0         29.0   
3    $30,299          3.2        6.0       270.0      23.0         28.0   
4    $39,014          3.5        6.0       225.0      21.0         24.0   
..        ...         ...        ...         ...       ...          ...   
427  $38,203          2.4        5.0       197.0      24.0         28.0   
428  $40,083          2.3        5.0       242.0      23.0         26.0   
429  $42,573          2.9        6.0       268.0      22.0         26.0   
430  $24,641          1.9        4.0       170.0      25.0         29.0   
431  $33,112          2.5        5.0       208.0      23.0         27.0   

     Weight  Wheelbase  Length  
0    4451.0      106.0   189.0  
1    2778.0      101.0   172.0  
2    3230.0      105.0   183.0  
3    3575.0      108.0   186.0  
4    3880.0      115.0   197.0  
..      ...        ...     ...  
427  3450.0      105.0   186.0  
428  3450.0      105.0   186.0  
429  3653.0      110.0   190.0  
430  2822.0      101.0   180.0  
431  3823.0      109.0   186.0  

[428 rows x 15 columns]